### PR TITLE
Moving from flat layout to src layout.

### PR DESCRIPTION
**Pull request by prady0t**
# Description

Moving to a src layout is a better option overall. You can read more about it here: https://packaging.python.org/en/latest/discussions/src-layout-vs-flat-layout/#:~:text=Auto%20color%20theme-,src%20layout%20vs%20flat%20layout,-%23

It was decided internally to move to a `src` layout. 

## Type of change

Please add a line in the relevant section of [CHANGELOG.md](https://github.com/pybamm-team/PyBaMM/blob/develop/CHANGELOG.md) to document the change (include PR #) - note reverse order of PR #s. If necessary, also add to the list of breaking changes.

- [ ] New feature (non-breaking change which adds functionality)
- [ ] Optimization (back-end change that speeds up the code)
- [ ] Bug fix (non-breaking change which fixes an issue)

# Key checklist:

- [ ] No style issues: `$ pre-commit run` (or `$ nox -s pre-commit`) (see [CONTRIBUTING.md](https://github.com/pybamm-team/PyBaMM/blob/develop/CONTRIBUTING.md#installing-and-using-pre-commit) for how to set this up to run automatically when committing locally, in just two lines of code)
- [ ] All tests pass: `$ python run-tests.py --all` (or `$ nox -s tests`)
- [ ] The documentation builds: `$ python run-tests.py --doctest` (or `$ nox -s doctests`)

You can run integration tests, unit tests, and doctests together at once, using `$ python run-tests.py --quick` (or `$ nox -s quick`).

## Further checks:

- [ ] Code is commented, particularly in hard-to-understand areas
- [ ] Tests added that prove fix is effective or that feature works

### CODE REVIEW

**Comment by prady0t** [comment]
@agriyakhetarpal I think I've added everything, please let me know if I have to add anything else here once the tests are completed.

File: pyproject.toml [hunk 1]
```diff
@@ -165,6 +165,7 @@ pybamm = [
 ]
 
 [tool.setuptools.packages.find]
+where = ["src"]
 include = ["pybamm", "pybamm.*"]
```
Comment (agriyakhetarpal) [inline pyproject.toml:169]: We might be able to remove this now and rely on `setuptools` to find the package under `src`

```suggestion
```

**Comment by agriyakhetarpal** [comment]
Just noticed, we'll have to fix the doctests and the example notebooks

**Comment by agriyakhetarpal** [comment]
Let's rebase this now, @prady0t

**Comment by prady0t** [comment]
@agriyakhetarpal  Opened a new PR, I wasn't able to fix the conflicts locally.

**Comment by agriyakhetarpal** [comment]
Superseded by #4311

**Pull request closed without merging**